[Rust] Entry API concept exercise (#1436)
exercism/v3: languages/rust/exercises/concept/entry-api/.docs/instructions.md

# Instructions

In this exercise you'll be using a `HashMap`, along with entry API methods, to solve a simple algorithm problem.

> The problem states: given a slice of `str`s representing a magazine and another slice of `str`s representing a note you're looking to construct, implement a function that returns a boolean indicating whether the note can be constructed from the magazine. In other words, does the magazine contain at least as many instances of each word required by the note?

Reviewer: discussion point: should we say `&str` here? Since `str` is also a type.

> The problem states: given a slice of `&str`s representing a magazine and another slice of `&str`s representing a note you're looking to construct, implement a function that returns a `bool` indicating whether the note can be constructed from the magazine. In other words, does the magazine contain at least as many instances of each word required by the note?

Reviewer 2: ```suggestion
Given `&[&str]` representing the words of a magazine article, and `&[&str]` representing the words of a note you would like to send, can you compose your note by cutting words out of the magazine and pasting them into a letter? 

Notes: 

- This implies physical cutting and pasting; the magazine needs to contain at least as many copies of each word as the note requires.
- Capitalization matters; just because you're pasting together a note composed from words of a magazine doesn't mean you're willing to be ungrammatical.
```

Rewrote this with the following goals:

- Put the problem statement first, not the programmatic solution to the problem
- Make problem boundaries explicit.

You'll start with the following stubbed function signature:

```rust

> pub fn construct_note(magazine: &[&str], note: &[&str]) -> bool {

Reviewer 2: ```suggestion
pub fn can_construct_note(magazine: &[&str], note: &[&str]) -> bool {
```

Naming boolean methods `can_*` or `is_*` improves legibility, and prevents future conflicts with a future method which actually does the work of constructing the desired note.

}
```

> Given the following input
> 
> ```rust
> let magazine = &["two", "times", "three", "is", "not", "four"];
> let note = &["two", "times", "two", "is", "four"];
> ```
> 
> the function should return `false` since the magazine only contains one instance of `"two"` when the note requires two of them.
> 
> Note that case sensitivity matters.

Reviewer 2: ```suggestion
    ```rust
    let magazine = "two times three is not four".split_whitespace();
    let note = "two times two is four".split_whitespace();
    assert_eq!(can_construct_note(magazine, note), false);
    ```
    
    The function returns `false` since the magazine only contains one instance of `"two"` when the note requires two of them.
```

This construction is shorter and more explicit. The note about capitalization has been moved to a previous section.

If doing this manually, please omit the leading spaces; those were only necessary to get github's suggestion box to work properly.
